src/mbgl/renderer/sources/render_image_source.cpp

```diff
@@ -114,44 +114,39 @@ void RenderImageSource::update(Immutable<style::Source::Impl> baseImpl_,
         return;
     }
 
-    auto size = transformState.getSize();
-    const double viewportHeight = size.height;
-
-    // Compute the screen coordinates at wrap=0 for the given LatLng
-    ScreenCoordinate nePixel = { -INFINITY, -INFINITY };
-    ScreenCoordinate swPixel = { INFINITY, INFINITY };
-
+    // Compute the z0 tile coordinates for the given LatLngs
+    TileCoordinatePoint nePoint = { -INFINITY, -INFINITY };
+    TileCoordinatePoint swPoint = { INFINITY, INFINITY };
+    std::vector<TileCoordinatePoint> tileCoordinates;
     for (LatLng latLng : coords) {
-        ScreenCoordinate pixel = transformState.latLngToScreenCoordinate(latLng);
-        swPixel.x = std::min(swPixel.x, pixel.x);
-        nePixel.x = std::max(nePixel.x, pixel.x);
-        swPixel.y = std::min(swPixel.y, viewportHeight - pixel.y);
-        nePixel.y = std::max(nePixel.y, viewportHeight - pixel.y);
-    }
-    const double width = nePixel.x - swPixel.x;
-    const double height = nePixel.y - swPixel.y;
+        auto point = TileCoordinate::fromLatLng(0, latLng).p;
+        tileCoordinates.push_back(point);
+        swPoint.x = std::min(swPoint.x, point.x);
+        nePoint.x = std::max(nePoint.x, point.x);
+        swPoint.y = std::min(swPoint.y, point.y);
+        nePoint.y = std::max(nePoint.y, point.y);
+   }
+
+    // Calculate the optimum zoom level to determine the tile ids to use for transforms
+    auto dx = nePoint.x - swPoint.x;
+    auto dy = nePoint.y - swPoint.y;
+    auto dMax = std::max(dx, dy);
+    double zoom = std::max(0.0, std::floor(-util::log2(dMax)));
 
     // Don't bother drawing the ImageSource unless it occupies >4 screen pixels
-    enabled = (width * height > 4);
+    enabled = dMax * std::pow(2.0, transformState.getZoom()) > 2.0 / util::tileSize;
```

Comment:
Number of pixels in a tile between the x or y extents of the tile coordinates at the current zoom level.

Reply:
I don't think I understand what's going on here, a comment in the code would be nice.

Here's how I'd think of it: scale factor is `std::pow(2.0, transformState.getZoom())`. Assuming scale of 1 (i.e. z0 tile), extent is 8192 and tileSize is 512, so you'd expect one tile unit to be 1/16th of a pixel, so 4 pixels would be 64 tile units. I must be missing some conversion factor used in `TileCoordinatePoint`?

Reply:
> a comment in the code would be nice.
👍

Your calculation is correct:
```
pixelsPerTileUnit = scaleFactor * util::EXTENT / util::tileSize;
```

But mine was not 😞 
`dMax` is in tile coordinates not `GeometryCoordinates`.  It needs to be scaled by `util::EXTENT`. The correct comparison would then be:
`(dMax * util::EXTENT) * (scaleFactor * util::EXTENT / util::tileSize) > 2 * util::EXTENT / util::tileSize`

This will enable image sources when they span >2px on their long side.

Reply:
> a comment in the code would be nice.
👍

At any zoom level 1 geometry coordinate unit would be `tileSize/extent` pixels.

`dMax` is in tile coordinate units at z0,  so we first scale it by `util::EXTENT` and then by  `2^z` (the scale factor) to get geometry coordinates at the current zoom.  The long side then becomes :`dMax * scale * extent * tileSize / extent` pixels.

This will enable image sources when they span >2px on their long side.